File: java/src/main/java/org/rocksdb/TypeUtil.java
// Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
//  This source code is licensed under both the GPLv2 (found in the
//  COPYING file in the root directory) and Apache 2.0 License
//  (found in the LICENSE.Apache file in the root directory).
package org.rocksdb;

import java.nio.ByteBuffer;

/**
 * Util class to get internal key for seeking from user key.
 */
public class TypeUtil {
  public static byte[] getInternalKey(final byte[] userKey, final Options options) {
    if (options == null || userKey == null) {
      throw new IllegalArgumentException("ByteBuffer and options parameters must not be null");
    }
    return getInternalKeyJni(userKey, userKey.length, options.getNativeHandle());
  }

  public static int getInternalKey(final ByteBuffer userKey, final ByteBuffer internalKey, final Options options) {
    int result;
    if (options == null || userKey == null || internalKey == null) {
      throw new IllegalArgumentException("ByteBuffer and options parameters must not be null");
    }
    if (userKey.isDirect()) {
      if (internalKey.isDirect()) {
        result =
            getInternalKeyDirect0(userKey, userKey.position(), userKey.remaining(), internalKey,
                internalKey.position(), internalKey.remaining(), options.getNativeHandle());
      } else {
        result = getInternalKeyDirect1(userKey, userKey.position(), userKey.remaining(),
            internalKey.array(), internalKey.arrayOffset() + internalKey.position(),
            internalKey.remaining(), options.getNativeHandle());
      }
    } else {
      if (internalKey.isDirect()) {
        result = getInternalKeyByteArray0(userKey.array(),
            userKey.arrayOffset() + userKey.position(), userKey.remaining(), internalKey,
            internalKey.position(), internalKey.remaining(), options.getNativeHandle());
      } else {
        result = getInternalKeyByteArray1(userKey.array(),
            userKey.arrayOffset() + userKey.position(), userKey.remaining(), internalKey.array(),
            internalKey.arrayOffset() + internalKey.position(), internalKey.remaining(),
            options.getNativeHandle());
      }
    }
    userKey.position(userKey.limit());
    internalKey.limit(Math.min(internalKey.position() + result, internalKey.limit()));
    return result;
  }

  public static byte[] getInternalKeyForPrev(final byte[] userKey, final Options options) {
    if (options == null || userKey == null) {
      throw new IllegalArgumentException("Byte array and options parameters must not be null");
    }
    return getInternalKeyForPrevJni(userKey, userKey.length, options.getNativeHandle());
  }

  public static int getInternalKeyForPrev(
      final ByteBuffer userKey, final ByteBuffer internalKey, final Options options) {
    if (options == null || userKey == null || internalKey == null) {
      throw new IllegalArgumentException("ByteBuffer and options parameters must not be null");
    }
    int result;
    if (userKey.isDirect()) {
      if (internalKey.isDirect()) {
        result = getInternalKeyDirectForPrev0(userKey, userKey.position(), userKey.remaining(),
            internalKey, internalKey.position(), internalKey.remaining(),
            options.getNativeHandle());
      } else {
        result = getInternalKeyDirectForPrev1(userKey, userKey.position(), userKey.remaining(),
            internalKey.array(), internalKey.arrayOffset() + internalKey.position(),
            internalKey.remaining(), options.getNativeHandle());
      }
    } else {
      if (internalKey.isDirect()) {
        result = getInternalKeyByteArrayForPrev0(userKey.array(),
            userKey.arrayOffset() + userKey.position(), userKey.remaining(), internalKey,
            internalKey.position(), internalKey.remaining(), options.getNativeHandle());
      } else {
        result = getInternalKeyByteArrayForPrev1(userKey.array(),
            userKey.arrayOffset() + userKey.position(), userKey.remaining(), internalKey.array(),
            internalKey.arrayOffset() + internalKey.position(), internalKey.remaining(),
            options.getNativeHandle());
      }
    }
    userKey.position(userKey.limit());
    internalKey.limit(Math.min(internalKey.position() + result, internalKey.limit()));
    return result;
  }

  private static native int getInternalKeyDirect0(final ByteBuffer userKey, final int userKeyOffset,
      int userKeyLen, ByteBuffer internalKey, int internalKeyOffset, int internalKeyLen,
      long optionsHandle);
  private static native int getInternalKeyByteArray0(final byte[] userKey, final int userKeyOffset,
      final int userKeyLen, final ByteBuffer internalKey, final int internalKeyOffset, final int internalKeyLen,
      final long optionsHandle);
  private static native int getInternalKeyDirect1(final ByteBuffer userKey, final int userKeyOffset,
      final int userKeyLen, final byte[] internalKey, final int internalKeyOffset, final int internalKeyLen,
      final long optionsHandle);
  private static native int getInternalKeyByteArray1(final byte[] userKey, final int userKeyOffset,
      final int userKeyLen, final byte[] internalKey, final int internalKeyOffset, final int internalKeyLen,
      final long optionsHandle);
  private static native byte[] getInternalKeyJni(
      final byte[] userKey, final int userKeyLen, final long optionsHandle);

  private static native int getInternalKeyDirectForPrev0(final ByteBuffer userKey, final int userKeyOffset,
      final int userKeyLen, final ByteBuffer internalKey, final int internalKeyOffset, final int internalKeyLen,
      final long optionsHandle);
  private static native int getInternalKeyByteArrayForPrev0(final byte[] userKey, final int userKeyOffset,
      final int userKeyLen, final ByteBuffer internalKey, final int internalKeyOffset, final int internalKeyLen,
      final long optionsHandle);
  private static native int getInternalKeyDirectForPrev1(final ByteBuffer userKey, final int userKeyOffset,
      final int userKeyLen, final byte[] internalKey, final int internalKeyOffset, final int internalKeyLen,
      final long optionsHandle);
  private static native int getInternalKeyByteArrayForPrev1(final byte[] userKey, final int userKeyOffset,
      final int userKeyLen, final byte[] internalKey, final int internalKeyOffset, final int internalKeyLen,
      final long optionsHandle);
  private static native byte[] getInternalKeyForPrevJni(
      final byte[] userKey, final int userKeyLen, final long optionsHandle);
}

File: java/rocksjni/type_util.cc
// Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
//  This source code is licensed under both the GPLv2 (found in the
//  COPYING file in the root directory) and Apache 2.0 License
//  (found in the LICENSE.Apache file in the root directory).
//
// This file implements the "bridge" between Java and C++ and enables
// calling c++ ROCKSDB_NAMESPACE::Iterator methods from Java side.

#include <jni.h>
#include <stdlib.h>

#include "include/org_rocksdb_TypeUtil.h"
#include "rocksdb/options.h"
#include "rocksdb/utilities/types_util.h"
#include "rocksjni/portal.h"

/*
 * Class:     org_rocksdb_TypeUtil
 * Method:    getInternalKeyDirect0
 * Signature: (Ljava/nio/ByteBuffer;IILjava/nio/ByteBuffer;II)I
 */
jint JNICALL Java_org_rocksdb_TypeUtil_getInternalKeyDirect0(
    JNIEnv *env, jclass /*clz*/, jobject user_key, jint user_key_off,
    jint user_key_len, jobject int_key, jint int_key_off, jint int_key_len,
    jlong options_handle) {
  auto *options =
      reinterpret_cast<const ROCKSDB_NAMESPACE::Options *>(options_handle);
  std::string seek_key_buf;
  auto getInternalKeySeek = [&seek_key_buf,
                             &options](ROCKSDB_NAMESPACE::Slice &target_slice) {
    ROCKSDB_NAMESPACE::GetInternalKeyForSeek(target_slice, options->comparator,
                                             &seek_key_buf);
  };
  ROCKSDB_NAMESPACE::JniUtil::k_op_direct(getInternalKeySeek, env, user_key,
                                          user_key_off, user_key_len);
  ROCKSDB_NAMESPACE::Slice key_slice = seek_key_buf;
  return ROCKSDB_NAMESPACE::JniUtil::copyToDirect(env, key_slice, int_key,
                                                  int_key_off, int_key_len);
}

/*
 * Class:     org_rocksdb_TypeUtil
 * Method:    getInternalKeyByteArray0
 * Signature: ([BIILjava/nio/ByteBuffer;IIJ)I
 */
jint JNICALL Java_org_rocksdb_TypeUtil_getInternalKeyByteArray0(
    JNIEnv *env, jclass /*clz*/, jbyteArray user_key, jint user_key_off,
    jint user_key_len, jobject int_key, jint int_key_off, jint int_key_len,
    jlong options_handle) {
  auto *options =
      reinterpret_cast<const ROCKSDB_NAMESPACE::Options *>(options_handle);
  std::string seek_key_buf;
  auto getInternalKeySeek = [&seek_key_buf,
                             &options](ROCKSDB_NAMESPACE::Slice &target_slice) {
    ROCKSDB_NAMESPACE::GetInternalKeyForSeek(target_slice, options->comparator,
                                             &seek_key_buf);
  };
  ROCKSDB_NAMESPACE::JniUtil::k_op_indirect(getInternalKeySeek, env, user_key,
                                            user_key_off, user_key_len);
  ROCKSDB_NAMESPACE::Slice key_slice = seek_key_buf;
  return ROCKSDB_NAMESPACE::JniUtil::copyToDirect(env, key_slice, int_key,
                                                  int_key_off, int_key_len);
}

/*
 * Class:     org_rocksdb_TypeUtil
 * Method:    getInternalKeyDirect1
 * Signature: (Ljava/nio/ByteBuffer;II[BIIJ)I
 */
jint JNICALL Java_org_rocksdb_TypeUtil_getInternalKeyDirect1(
    JNIEnv *env, jclass /*cls*/, jobject user_key, jint user_key_off,
    jint user_key_len, jbyteArray int_key, jint int_key_off, jint int_key_len,
    jlong options_handle) {
  auto *options =
      reinterpret_cast<const ROCKSDB_NAMESPACE::Options *>(options_handle);
  std::string seek_key_buf;
  auto getInternalKeySeek = [&seek_key_buf,
                             &options](ROCKSDB_NAMESPACE::Slice &target_slice) {
    ROCKSDB_NAMESPACE::GetInternalKeyForSeek(target_slice, options->comparator,
                                             &seek_key_buf);
  };
  ROCKSDB_NAMESPACE::JniUtil::k_op_direct(getInternalKeySeek, env, user_key,
                                          user_key_off, user_key_len);
  ROCKSDB_NAMESPACE::Slice key_slice = seek_key_buf;
  return ROCKSDB_NAMESPACE::JniUtil::copyToByteArray(env, key_slice, int_key,
                                                     int_key_off, int_key_len);
}

/*
 * Class:     org_rocksdb_TypeUtil
 * Method:    getInternalKeyByteArray1
 * Signature: ([BII[BIIJ)I
 */
jint JNICALL Java_org_rocksdb_TypeUtil_getInternalKeyByteArray1(
    JNIEnv *env, jclass /*cls*/, jbyteArray user_key, jint user_key_off,
    jint user_key_len, jbyteArray int_key, jint int_key_off, jint int_key_len,
    jlong options_handle) {
  auto *options =
      reinterpret_cast<const ROCKSDB_NAMESPACE::Options *>(options_handle);

  std::string seek_key_buf;
  auto getInternalKeySeek = [&seek_key_buf,
                             &options](ROCKSDB_NAMESPACE::Slice &target_slice) {
    ROCKSDB_NAMESPACE::GetInternalKeyForSeek(target_slice, options->comparator,
                                             &seek_key_buf);
  };
  ROCKSDB_NAMESPACE::JniUtil::k_op_indirect(getInternalKeySeek, env, user_key,
                                            user_key_off, user_key_len);
  ROCKSDB_NAMESPACE::Slice key_slice = seek_key_buf;
  return ROCKSDB_NAMESPACE::JniUtil::copyToByteArray(env, key_slice, int_key,
                                                     int_key_off, int_key_len);
}

/*
 * Class:     org_rocksdb_TypeUtil
 * Method:    getInternalKeyJni
 * Signature: ([BIJ)[B
 */
jbyteArray JNICALL Java_org_rocksdb_TypeUtil_getInternalKeyJni(
    JNIEnv *env, jclass /*cls*/, jbyteArray user_key, jint user_key_len,
    jlong options_handle) {
  jbyte *target = env->GetByteArrayElements(user_key, nullptr);
  if (target == nullptr) {
    // exception thrown: OutOfMemoryError
    return nullptr;
  }
  auto *options =
      reinterpret_cast<const ROCKSDB_NAMESPACE::Options *>(options_handle);
  ROCKSDB_NAMESPACE::Slice target_slice(reinterpret_cast<char *>(target),
                                        user_key_len);
  std::string seek_key_buf;
  ROCKSDB_NAMESPACE::GetInternalKeyForSeek(target_slice, options->comparator,
                                           &seek_key_buf);
  ROCKSDB_NAMESPACE::Slice key_slice = seek_key_buf;
  jbyteArray jkey = env->NewByteArray(static_cast<jsize>(key_slice.size()));
  if (jkey == nullptr) {
    ROCKSDB_NAMESPACE::OutOfMemoryErrorJni::ThrowNew(env, "Memory allocation failed in RocksDB JNI function");
    return nullptr;
  }
  ROCKSDB_NAMESPACE::JniUtil::copyToByteArray(
      env, key_slice, jkey, 0, static_cast<jint>(key_slice.size()));
  return jkey;
}

/*
 * Class:     org_rocksdb_TypeUtil
 * Method:    getInternalKeyDirectForPrev0
 * Signature: (Ljava/nio/ByteBuffer;IILjava/nio/ByteBuffer;IIJ)I
 */
jint JNICALL Java_org_rocksdb_TypeUtil_getInternalKeyDirectForPrev0(
    JNIEnv *env, jclass /*clz*/, jobject user_key, jint user_key_off,
    jint user_key_len, jobject int_key, jint int_key_off, jint int_key_len,
    jlong options_handle) {
  auto *options =
      reinterpret_cast<const ROCKSDB_NAMESPACE::Options *>(options_handle);
  std::string seek_key_buf;
  auto getInternalKeySeek = [&seek_key_buf,
                             &options](ROCKSDB_NAMESPACE::Slice &target_slice) {
    ROCKSDB_NAMESPACE::GetInternalKeyForSeekForPrev(
        target_slice, options->comparator, &seek_key_buf);
  };
  ROCKSDB_NAMESPACE::JniUtil::k_op_direct(getInternalKeySeek, env, user_key,
                                          user_key_off, user_key_len);
  ROCKSDB_NAMESPACE::Slice key_slice = seek_key_buf;
  return ROCKSDB_NAMESPACE::JniUtil::copyToDirect(env, key_slice, int_key,
                                                  int_key_off, int_key_len);
}

/*
 * Class:     org_rocksdb_TypeUtil
 * Method:    getInternalKeyByteArrayForPrev0
 * Signature: ([BIILjava/nio/ByteBuffer;IIJ)I
 */
jint JNICALL Java_org_rocksdb_TypeUtil_getInternalKeyByteArrayForPrev0(
    JNIEnv *env, jclass /*clz*/, jbyteArray user_key, jint user_key_off,
    jint user_key_len, jobject int_key, jint int_key_off, jint int_key_len,
    jlong options_handle) {
  auto *options =
      reinterpret_cast<const ROCKSDB_NAMESPACE::Options *>(options_handle);

  std::string seek_key_buf;
  auto getInternalKeySeek = [&seek_key_buf,
                             &options](ROCKSDB_NAMESPACE::Slice &target_slice) {
    ROCKSDB_NAMESPACE::GetInternalKeyForSeekForPrev(
        target_slice, options->comparator, &seek_key_buf);
  };
  ROCKSDB_NAMESPACE::JniUtil::k_op_indirect(getInternalKeySeek, env, user_key,
                                            user_key_off, user_key_len);
  ROCKSDB_NAMESPACE::Slice key_slice = seek_key_buf;
  return ROCKSDB_NAMESPACE::JniUtil::copyToDirect(env, key_slice, int_key,
                                                  int_key_off, int_key_len);
}

/*
 * Class:     org_rocksdb_TypeUtil
 * Method:    getInternalKeyDirectForPrev1
 * Signature: (Ljava/nio/ByteBuffer;II[BIIJ)I
 */
jint JNICALL Java_org_rocksdb_TypeUtil_getInternalKeyDirectForPrev1(
    JNIEnv *env, jclass /*cls*/, jobject user_key, jint user_key_off,
    jint user_key_len, jbyteArray int_key, jint int_key_off, jint int_key_len,
    jlong options_handle) {
  auto *options =
      reinterpret_cast<const ROCKSDB_NAMESPACE::Options *>(options_handle);
  std::string seek_key_buf;
  auto getInternalKeySeek = [&seek_key_buf,
                             &options](ROCKSDB_NAMESPACE::Slice &target_slice) {
    ROCKSDB_NAMESPACE::GetInternalKeyForSeekForPrev(
        target_slice, options->comparator, &seek_key_buf);
  };
  ROCKSDB_NAMESPACE::JniUtil::k_op_direct(getInternalKeySeek, env, user_key,
                                          user_key_off, user_key_len);
  ROCKSDB_NAMESPACE::Slice key_slice = seek_key_buf;
  return ROCKSDB_NAMESPACE::JniUtil::copyToByteArray(env, key_slice, int_key,
                                                     int_key_off, int_key_len);
}

/*
 * Class:     org_rocksdb_TypeUtil
 * Method:    getInternalKeyByteArrayForPrev1
 * Signature: ([BII[BIIJ)I
 */
jint JNICALL Java_org_rocksdb_TypeUtil_getInternalKeyByteArrayForPrev1(
    JNIEnv *env, jclass /*cls*/, jbyteArray user_key, jint user_key_off,
    jint user_key_len, jbyteArray int_key, jint int_key_off, jint int_key_len,
    jlong options_handle) {
  auto *options =
      reinterpret_cast<const ROCKSDB_NAMESPACE::Options *>(options_handle);
  std::string seek_key_buf;
  auto getInternalKeySeek = [&seek_key_buf,
                             &options](ROCKSDB_NAMESPACE::Slice &target_slice) {
    ROCKSDB_NAMESPACE::GetInternalKeyForSeekForPrev(
        target_slice, options->comparator, &seek_key_buf);
  };
  ROCKSDB_NAMESPACE::JniUtil::k_op_indirect(getInternalKeySeek, env, user_key,
                                            user_key_off, user_key_len);
  ROCKSDB_NAMESPACE::Slice key_slice = seek_key_buf;
  return ROCKSDB_NAMESPACE::JniUtil::copyToByteArray(env, key_slice, int_key,
                                                     int_key_off, int_key_len);
}

/*
 * Class:     org_rocksdb_TypeUtil
 * Method:    getInternalKeyForPrevJni
 * Signature: ([BIJ)[B
 */
jbyteArray JNICALL Java_org_rocksdb_TypeUtil_getInternalKeyForPrevJni(
    JNIEnv *env, jclass /*cls*/, jbyteArray user_key, jint user_key_len,
    jlong options_handle) {
  jbyte *target = env->GetByteArrayElements(user_key, nullptr);
  if (target == nullptr) {
    // exception thrown: OutOfMemoryError
    return nullptr;
  }
  auto *options =
      reinterpret_cast<const ROCKSDB_NAMESPACE::Options *>(options_handle);
  ROCKSDB_NAMESPACE::Slice target_slice(reinterpret_cast<char *>(target),
                                        user_key_len);
  std::string seek_key_buf;
  ROCKSDB_NAMESPACE::GetInternalKeyForSeekForPrev(
      target_slice, options->comparator, &seek_key_buf);
  ROCKSDB_NAMESPACE::Slice key_slice = seek_key_buf;
  jbyteArray jkey = env->NewByteArray(static_cast<jsize>(key_slice.size()));
  if (jkey == nullptr) {
    ROCKSDB_NAMESPACE::OutOfMemoryErrorJni::ThrowNew(env, "Memory allocation failed in RocksDB JNI function");
    return nullptr;
  }
  ROCKSDB_NAMESPACE::JniUtil::copyToByteArray(
      env, key_slice, jkey, 0, static_cast<jint>(key_slice.size()));
  return jkey;
}
